go/libraries/doltcore/sqle/statspro/jobqueue/serialqueue.go

```diff
@@ -0,0 +1,408 @@
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Run a normal priority job asynchronously on the queue. Returns once the
+// job is accepted.
+func (s *SerialQueue) DoAsync(f func() error) error {
+	_, err := s.submitWork(schedPriority_Normal, f)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Helper function to submit work. Returns the work submitted, if it
+// was successful, and an error otherwise.
+func (s *SerialQueue) submitWork(pri schedPriority, f func() error) (work, error) {
+	w := work{
+		f:    f,
+		done: make(chan struct{}),
+	}
+	err := s.makeReq(schedReq{
+		reqType: schedReqType_Enqueue,
+		pri:     pri,
+		work:    w,
+		resp:    make(chan schedResp, 1),
+	})
+	if err != nil {
+		return work{}, err
+	}
+	return w, nil
+}
+
+func (s *SerialQueue) makeReq(req schedReq) error {
+	select {
+	case s.schedCh <- req:
+		resp := <-req.resp
+		return resp.err
+	case <-s.completed:
+		return ErrCompletedQueue
+	}
+}
+
+// Read off the input channels and maintain queues of pending work.
+// Deliver that work to the runner channel if it is desired.
+func (s *SerialQueue) runScheduler(ctx context.Context) {
+	state := schedState_Running
+	normalQ := circular.NewBuff[work](16)
+	highQ := circular.NewBuff[work](16)
+	for {
+		var sendWorkCh chan work
+		var sendWork work
+		var sentWorkCallback func()
+
+		if state == schedState_Running {
+			if highQ.Len() > 0 {
+				sendWorkCh = s.runnerCh
+				sendWork = highQ.Front()
+				sentWorkCallback = highQ.Pop
+			} else if normalQ.Len() > 0 {
+				sendWorkCh = s.runnerCh
+				sendWork = normalQ.Front()
+				sentWorkCallback = normalQ.Pop
+			}
+		}
+
+		select {
+		case msg := <-s.schedCh:
+			switch msg.reqType {
+			case schedReqType_Enqueue:
+				if state == schedState_Stopped {
+					msg.resp <- schedResp{
+						err: ErrStoppedQueue,
+					}
+				} else {
+					if msg.pri == schedPriority_High {
+						highQ.Push(msg.work)
+					} else {
+						normalQ.Push(msg.work)
+					}
+					msg.resp <- schedResp{
+						err: nil,
+					}
+				}
+			case schedReqType_Purge:
+				highQ = circular.NewBuff[work](highQ.Cap())
+				normalQ = circular.NewBuff[work](normalQ.Cap())
+				msg.resp <- schedResp{
+					err: nil,
+				}
+			case schedReqType_Start:
+				state = schedState_Running
+				msg.resp <- schedResp{
+					err: nil,
+				}
+			case schedReqType_Pause:
+				state = schedState_Paused
+				msg.resp <- schedResp{
+					err: nil,
+				}
+			case schedReqType_Stop:
+				state = schedState_Stopped
+				msg.resp <- schedResp{
+					err: nil,
+				}
+			}
+		case sendWorkCh <- sendWork:
+			// Pop from queue the work came from.
+			sentWorkCallback()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// Read off the runner channel and run the submitted work.
+func (s *SerialQueue) runRunner(ctx context.Context) {
+	ticker := time.NewTicker(1)
+	for {
+		select {
+		case w := <-s.runnerCh:
+			if w.newRate > 0 {
+				ticker.Reset(w.newRate)
+			}
+			select {
```

Comment:
What's this select doing?

Reply:
can't execute stats jobs more frequently than the ticker rate